packages/core/src/common/messaging/connection-error-handler.ts

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { Message } from "vscode-jsonrpc";
+
+export interface ResolvedConnectionErrorHandlerOptions {
+    readonly serverName: string
+    /**
+     * The maximum amout of errors allowed before stopping the server.
+     */
+    readonly maxErrors: number
+    /**
+     * The maimum amount of restarts allowed in the restart interval.
+     */
+    readonly maxRestarts: number
+    /**
+     * In minutes.
+     */
+    readonly restartInterval: number
+}
+
+export type ConnectionErrorHandlerOptions = Partial<ResolvedConnectionErrorHandlerOptions> & {
+    readonly serverName: string
+};
+
+export class ConnectionErrorHandler {
+
+    protected readonly options: ResolvedConnectionErrorHandlerOptions;
+    constructor(options: ConnectionErrorHandlerOptions) {
+        this.options = {
+            maxErrors: 3,
+            maxRestarts: 5,
+            restartInterval: 3,
+            ...options
+        };
+    }
+
+    shouldStop(error: Error, message?: Message, count?: number): boolean {
+        return !count || count > this.options.maxErrors;
+    }
+
+    protected readonly restarts: number[] = [];
+    shouldRestart(): string | undefined {
```

Comment:
Seems this should be shouldResart( { reason: string }): boolean ?

Reply:
Or I'm thinking in C too much and it shoulb be shouldResart() : { success: boolean, reason: string }

Reply:
I've changed with logging in the error handler.